llap-server/src/java/org/apache/hadoop/hive/llap/io/api/impl/LlapIoImpl.java

```diff
@@ -221,6 +227,9 @@ public void debugDumpShort(StringBuilder sb) {
         metadataCache, dataCache, bufferManagerOrc, conf, cacheMetrics, ioMetrics, tracePool);
     this.genericCvp = isEncodeEnabled ? new GenericColumnVectorProducer(
         serdeCache, bufferManagerGeneric, conf, cacheMetrics, ioMetrics, tracePool) : null;
+    proactiveEvictionExecutor = Executors.newSingleThreadExecutor(
```

Comment:
I guess I just wanted to make sure that this operation is completely decoupled from the rest of threads.
On daemon side you're right we'd be in an RPC handler thread while doing this eviction. My idea was that this request should return as fast as possible and don't even interfere with the RPC threads. Hence the void return value btw.

On HS2 side in ProactiveEviction we already use an executor service to send the eviction request. 
 So it's decoupled on HS2 too. If you feel that's enough I'll remove the additional async part on the daemon side.

Reply:
@szlta , if for every new feature we add a new thread pool then the code will become a huge burden to maintain. The RPC thread pool is meant for that doing the stuff within  thread of the pool that can be increased if needed. By creating a new pool we are not changing anything just adding more burden from code perspective.  Same on the HS2I user will need to wait and see the result of the command same with the purge.

Reply:
Well I agree on the daemon side, but not on iHS2. I think this should not handled the same way purge is:

- the invocation of purge is done by the users themselves so they should know what they're doing and this can wait for the result however long it takes
- proactive eviction is done in the background i.e. less apparent to the users - e.g. whenever a drop table/partition/db happens. I'd imagine a use case where we have many many daemons with large caches, and such proactive eviction could take longer. In this case I don't want users having to wait for the eviction to complete before their drop table command returns.
Wouldn't you agree?

Reply:
I do not agree, When you say wait can you explain what that means ?
Whom is waiting and where and why? 
Then how spinning up new thread pool will help that if there is contention?
Also you are mentioning HS2I but this code is in the LLAP runner so not sure how to read this.
Finally when you say long time or wait how long you think will take ? i do not see how this is long and do not see what it is blocking.

Reply:
Like I said, I'm fine with relying on RPC threadpool only on the LLAP daemon side. I'm not talking about that anymore (you can see I amended that change already)
What I'm talking about now is HS2 side:
- purge currently will block user side: https://github.com/szlta/hive/blob/HIVE-22821/ql/src/java/org/apache/hadoop/hive/ql/processors/LlapCacheResourceProcessor.java#L141
- while proactive eviction - I'd propose - should not: https://github.com/szlta/hive/blob/HIVE-22821/ql/src/java/org/apache/hadoop/hive/llap/ProactiveEviction.java#L79

That's the difference I'm interested in.

Reply:
I do think you see my point,  by adding this new pool what are you actually unlocking/unclocking ? 
Which Threads/actors are blocked and how this is helping?
When you say HS2, can you explain what is blocked there and how that have an impact on the system ?
What am trying to say is that there is nothing to really unblock and this is adding some code that is not needed IMO.

Reply:
Okay, so suppose you use beeline and do a drop table statement:
A thread from HiveServer2-Handler-Pool will pick up the incoming Thrift request and execute. 
Execution will go into DropTableOperation. Anything happening here will prevent this thread from finishing and thus you as the user on the beeline side (as thrift connects you) will have to wait.

If we invoke a proactive eviction from DropTableOperation (which this change does) obviously the user on beeline side will see this addition processing time for their query.
Now if this additional time is long - that could be bad, since they only wanted to drop a table and now they're waiting extra seconds if the DropTableOperation 'waits' for many LLAP daemons to finish their eviction on their big cache structure.

Reply:
Again this is a premature Optimization, evicting from the cache is walking a list of couple of millions items at most, that is a fraction of second. Then if this is a real issue it would be better to deal with it within HS2 by making this fire and forget kind of request. 
Again i do not see why this is needed and why it is done on the llap side.

Reply:
The other element that this will entail as well is that will have 2 paths one for the case the user has issue  drop table form the cache and the second when the table is dropped via command. IMO when to tackle the first one, where the user is actually issuing a command to explicitly drop the entries form the cache. The second one is a bonus..

Reply:
"better to deal with it within HS2 by making this fire and forget kind of request"
This is exactly my intention and how the code is right now, as you can see this if you read back and take a look on the code.

"fraction of second"
For one daemon, yes - what if we have many? And how will the remote invocations' overhead add up to this?

"why it is done on the llap side"
I already stated 2 times that I'm fine and agree with you on not doing anything additional async stuff on LLAP daemon side, btw did you not see this commit in this PR?
https://github.com/apache/hive/pull/909/commits/ebf39c472dd311e9b7b3d93061736fcac3e20240